apps/frontend-pwa/src/pages/course/[courseId]/element/[id].tsx

```diff
@@ -89,18 +90,28 @@ function LearningElement({ courseId, id }: Props) {
       variables: {
         courseId: router.query.courseId as string,
         id: currentInstance?.id as number,
-        response:
-          questionData?.type === QuestionType.SC ||
-          questionData?.type === QuestionType.MC ||
-          questionData?.type === QuestionType.KPRIM
-            ? {
-                choices: response as number[],
-              }
-            : { value: response as string },
+        response: formatResponse(),
       },
     })
   }
 
+  const formatResponse = () => {
+    if (
+      questionData?.type === QuestionType.SC ||
+      questionData?.type === QuestionType.MC
+    ) {
+      return { choices: response as number[] }
+    } else if (questionData?.type === QuestionType.KPRIM) {
+      let choicesArray: number[] = []
+      Object.keys(response).map((key) => {
```

Comment:
refactor so it does not use let and array.push (mutable way)

-> naive approach: .filter and then .map, store the returned value of .map in choicesarray
-> better approach: use .flatMap to combine .filter and .map (return [] in flatMap to skip the element, [value] to add it to the output). flatMap combines all of the returned arrays into one large array, so empty arrays -> filtered out

Reply:
also, add questionData and response etc. as an argument to formatResponse and place it as a function outside of the react component (cleaner and better reusability). best approach would be a custom hook, but not necessary here

Reply:
thanks for the feedback! I adjusted the code accordingly 👍